### PR TITLE
Add ItemTemplates for Razor Pages

### PR DESCRIPTION
**Pull request by rynowak**
/cc @damianedwards @mlorbetske

### CODE REVIEW

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
+  "author": "Microsoft",
+  "classifications": ["Web", "ASP.NET"],
+  "name": "Razor Page with Page Model",
```
Comment (rynowak) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json]: @damianedwards @eilon - do the branding thing plz

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  "generatorVersions": "[1.0.0.0-*)",
+  "groupIdentity": "Microsoft.AspNetCore.Mvc.RazorPageWithPageModel",
+  "precedence": "100",
+  "identity": "Microsoft.AspNetCore.Mvc.RazorPageWithPageModel",
+  "shortName": "page",
```
Comment (rynowak) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json]: Ideas about a short name to distinguish this from pages (without pagemodel)??

We also have the option to make this a single item and have another parameter to distinguish.

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/Index.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+
+namespace MyApp.Namespace
```
Comment (rynowak) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/Index.cs]: This does the right thing. Relax.

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json (hunk 2 above)
Comment (seancpeters) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json]: I'd advocate for making these into a single template with a distinguishing param, maybe something like:
```
"model": {
  "type": "parameter",
  "datatype": "bool",
  "defaultValue": "false",
  "description": "<insert appropriate desc. here>"
}
}

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  "sourceName": "Index",
+  "primaryOutputs": [ { "path": "Index.cshtml"}, { "path": "Index.cs" } ],
+  "defaultName": "Index",
+  "symbols": {
+    "namespace": {
```
Comment (seancpeters) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json]: It's probably good to add a description to the namespace symbol, it'll get displayed in the detailed help for the template in the CLI.

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json (hunk 1 above)
Comment (Eilon) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json]: Not it.

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/Index.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.RazorPages;
```
Comment (Eilon) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/Index.cs]: Is this a good default set of namespaces? Anything else common that should go here? (Not at all saying it isn't, I'm just not familiar with what's generally used.)

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/Index.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace MyApp.Namespace
+{
+    public class Index : PageModel
+    {
+        public void OnGet()
+        {
+
```
Comment (Eilon) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/Index.cs]: Whitespace? Is this a joke?

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/Index.cs (hunk 5 above)
Comment (seancpeters) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/Index.cs]: For templates, we generally only include the usings that are required by the generated template code. But I could see including additional usings that most users would end up adding as they modify the content.

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json (hunk 2 above)
Comment (rynowak) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json]: I'll give this a try thanks

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json (hunk 4 above)
Comment (rynowak) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/.template.config/template.json]: 👍

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/Index.cs (hunk 5 above)
Comment (rynowak) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/Index.cs]: This is the same set of namespaces that the template uses for `HomeController` with the `Microsoft.AspNetCore.Mvc.RazorPages` namespace added (required).

Comment (Eilon) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPageWithPageModel/Index.cs]: This looks good to me then.

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/.template.config/template.json [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  "precedence": "100",
+  "identity": "Microsoft.AspNetCore.Mvc.RazorPage",
+  "shortName": "page",
+  "sourceName": "Index",
+  "primaryOutputs": [ { "path": "Index.cshtml"}, { "path": "Index.cs" } ],
+  "defaultName": "Index",
+  "sources": {
+    "modifiers": [
+      {
+        "condition": "(no-pagemodel)",
```
Comment (rynowak) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/.template.config/template.json]: @seancpeters - this isn't working. I've tried some stuff and it still wants to output the `Index.cs` file.

Comment (seancpeters) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/.template.config/template.json]: I screwed up the syntax when we were chatting. Sources should be like this:
```
  "sources": [
    {
      "modifiers": [
        {
          "condition": "(no-pagemodel)",
          "exclude": [
            "Index.cs"
          ]
        }
      ]
    }
  ],
```

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/Index.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+
+namespace MyApp.Namespace
+{
+    public class Index : PageModel
```
Comment (DamianEdwards) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/Index.cs]: Pls name this `IndexModel` and the file `Index.cshtml.cs`

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/Index.cshtml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+﻿@page
+@*#if (!no-pagemodel)*@
+@model Index
+@*#endif*@
+@{
+}
```
Comment (DamianEdwards) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/Index.cshtml]: Don't​we usually include blank lines at the end of template files? I think it makes it easier to start adding HTML, etc.

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/.template.config/template.json [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        "exclude": [
+          "Index.cs"
+        ]
+      }
+    ]
+  },
+  "symbols": {
+    "namespace": {
```
Comment (seancpeters) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/.template.config/template.json]: @rynowak - you had asked me about making a parameter required. You can add 
`"isRequired": "true" `
... to the symbol definition to make it required.

Note: I'm tracking down a bug related to isRequired, so you may want to hold off on making that change.

Comment (rynowak) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/.template.config/template.json]: Sure, I wasn't planning to do it for this change, I just wanted to know.

**Comment by rynowak** [comment]
> Should we add item templates for _ViewStart.cshmtl and _ViewImports.cshtml too?

Sure, I'll do this in a follow up PR

**Comment by rynowak** [comment]
```
dotnet new page
```
Generates Page and PageModel, uses a placeholder namespace

----------------------

```
dotnet new page --namespace MyCoolApp.MyCoolNamespace
```
Generates Page and PageModel, uses provided namespace

--------------------------

```
dotnet new page --no-pagemodel
```

Generates Page

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/Index.cs (hunk 8 above)
Comment (rynowak) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/Index.cs]: 👍

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/Index.cshtml (hunk 9 above)
Comment (rynowak) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/Index.cshtml]: I don't usually know. Added a blank line

File: template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/.template.config/template.json [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  "author": "Microsoft",
+  "classifications": ["Web", "ASP.NET"],
+  "name": "Razor Page",
+  "generatorVersions": "[1.0.0.0-*)",
+  "groupIdentity": "Microsoft.AspNetCore.Mvc.RazorPage",
+  "precedence": "100",
+  "identity": "Microsoft.AspNetCore.Mvc.RazorPage",
+  "shortName": "page",
+  "sourceName": "Index",
+  "primaryOutputs": [ { "path": "Index.cshtml"}, { "path": "Index.cs" } ],
```
Comment (seancpeters) [inline template_feed/Microsoft.DotNet.Common.ItemTemplates/content/RazorPage/.template.config/template.json]: @rynowak - for adding conditions to primary outputs, the format is like this:
```
"primaryOutputs": [ 
  {  "path": "Index.cshtml" }, 
  { 
    "path": "Index.cs",
    "condition": "(!no-pagemodel)"
  } 
]
```
If I understood correctly, that example should do what we talked about earlier.

**Comment by rynowak** [comment]
Updated

**Comment by rynowak** [comment]
@mlorbetske @seancpeters - Ping. this is ready to go

**Pull request merged**